### DISCUSSION TITLE
How is a compartment defined in arbor.cell_kind.cable?

### DISCUSSION THREAD

**Discussion by lungsi** [arbor-sim/arbor#1240] [Q&A] [answered]
`Arbor Version: 0.4.1.dev0`

Given that
- a *segment* is an arbor tree segment
	* a *region* may be composed of > 1 segment (with the same `tag`)
- a *cable* is created from the tree
- the *morphology* is the cell geometry as branched tree
	* DEFAULT: 1 compartment per branch

### What does this mean?
If a branch is composed of more than one segment and not all segments belong to the same region, for example branch 0 (right figure shown below)
![Example morphology](https://arbor.readthedocs.io/en/latest/_images/label_morph.svg)
**Does this mean that regardless of what region the segments might be part of, all the segments within the branch are considered a single compartment?**
![Example compartment](http://upload.wikimedia.org/wikipedia/commons/thumb/3/33/MembraneCircuit.jpg/500px-MembraneCircuit.jpg)

# Questions on functionalities to discretize compartments.

## 1. `cell.compartments_per_branch(n)` such that *n* &isin; *&#8484;+*
   This function is self-explnatory.

### What is `maxlen`? and What does the function do?

Instead of the default 1 compartment per branch if this discretizes the compartment based on *&#8467;* &mu;m **Will the new n-compartments be such that each are &#8467; &mu;m long?**

If this is the case, what happens if the branch cannot have integral divisions, i.e, (branch length) &divide; &#8467; = n &notin; &#8484;+

### What does this function do?
Doing `help(cell.compartments_on_segments)` returned
```
Help on method compartments_on_segments in module arbor._arbor:

compartments_on_segments(...) method of arbor._arbor.cable_cell instance
    compartments_on_segments(self: arbor._arbor.cable_cell) -> None
 
    Decompose each branch into compartments defined by segments.
```
The function appears to only take its instantiated self as the argument.

**Comment by bcumming** [comment, marked as answer]
Arbor makes a very deliberate distinction between the _description_ of a morphology  (segments, branches, samples), and the _discretisation_ used by the numerical solver (control volumes, roughly analogous to compartments in NEURON).

The [morphology description](https://arbor.readthedocs.io/en/latest/concepts/morphology.html), using segments and branches is used to describe a cell shape. [Regions and locations](https://arbor.readthedocs.io/en/latest/concepts/labels.html) on the cell refer to this description, and are used to describe where mechanisms, probes, stims, etc are located on the morphology.

The process of decomposing the morphology into control volumes (CVs) is performed during model construction. How the morphology is to be decomposed is governed by a _cv policy_, which is a set of rules that describe how the branches.

**note** NEURON users will be familiar with the term "compartment", which is an unbranched cable section with fixed radius that is used . Arbor's equivalent concept is the control volume (CV). Unlike NEURON, CVs can have more complicated geometry, for example the radius can vary in a CV, and CVs can even include fork points and sub-trees of the morphology. For this reason we use the term CV instead of compartment, to avoid confusion (we are in the process of purging the word compartment from the docs).

**warning** the Python interface in v0.4 for setting cv policies is very limited, and will change significantly in v0.5. Currently the API allows you to set a single policy (e.g. `compartments_per_branch` or `compartments_length`) on the whole cell. This will change to allow you to set different rules on different parts of the morphology, e.g. "use one CV for the soma, have one CV per branch in the apical dendrites, make CVs with maximum length 10 um on the dendrites". **I will update this answer to refer to the new API as soon as it is merged**.

`cell.compartments_length(L)`: break each branch into `N` CVs of equal length, where `N` is chosen to be the smallest number that ensures no CV will have a length on the branch longer than `L` micrometres. So, for example, a branch of length 15 um would be broken into `N=2` CVs of length 7.5 um when `L=10`. A branch of length 30um would be decomposed into 3 CVs of length 10 um, and a branch of length 5 um would be decomposed into a single CV of length 5 um.

`cell.compartments_on_segments()`: does not take any additional arguments, because it creates one compartment for every segment in the original morphology description. This is useful if you want, for example, one compartment for every segment between samples in an input SWC file, or for every segment in an input neuroml file.

**Comment by brenthuisman** [comment]
The changes announced in @bcumming 's message have been merged, together with (hopefully) more carefully defined and thus clear definitions of the various aspects of cable cell morphology, including control volumes.

Please see:
- https://arbor.readthedocs.io/en/latest/concepts/morphology.html
- https://arbor.readthedocs.io/en/latest/concepts/cable_cell.html#discretisation-and-cv-policies

@lungsi Does this help clear up your questions? If not, please get back to us, thanks!
